### PR TITLE
Add file search utilites

### PR DESCRIPTION
**Pull request by duehoa1211**
Link to #731

### CODE REVIEW

**Comment by ElementalCrisis** [comment]
Looking good! Just a couple issues. 

1. Search doesn't seem to work. if I type in "One" I get no results. 
https://github.com/ShokoAnime/Shoko-WebUI/assets/9443295/0e835737-327c-4211-95b0-fcc4aab06124

2. Selected Files bleed through the header when scrolling.
image

There are some issues with the selected panel but that's more UI issues I can play around with later.

**Comment by duehoa1211** [comment]
> Looking good! Just a couple issues.
> 
> 1. Search doesn't seem to work. if I type in "One" I get no results when One Piece files should appear.
> 2. Selected Files bleed through the header when scrolling.
> 
> There are some issues with the selected panel but that's more UI issues I can play around with later.

I have fixed 
> 2. Selected Files bleed through the header when scrolling.the selected files bleed

but for the
> 1. Search doesn't seem to work. if I type in "One" I get no results when One Piece files should appear.

i think this one is back-end problem since the endpoint has changed.

**Comment by harshithmohan** [comment]
Yeah... Someone reported it wasn't working in the unrecog util also.

Sent with Shortwave <https://www.shortwave.com?utm_medium=email&utm_content=signature&utm_source=Z2hhcnNoaXRobW9oYW5AZ21haWwuY29t>

File: src/core/react-query/episode/queries.ts [hunk 1]
```diff
@@ -18,3 +19,10 @@ export const useEpisodeFilesQuery = (
     select: transformListResultSimplified,
     enabled,
   });
+
+export const useEpisodeAniDBQuery = (episodeId: number, enabled = true) =>
+  useQuery<EpisodeAniDBType>({
+    queryKey: ['episode', 'single', episodeId],
```
Comment (harshithmohan) [inline src/core/react-query/episode/queries.ts:25]:
```suggestion
    queryKey: ['episode', 'anidb', episodeId],
```

File: src/pages/utilities/FileSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+      />
+    </div>
+  );
+};
+
+const FileSelected = (props: FileSelectedProps) => {
+  const { fileId } = props;
+  const { data: file } = useFileQuery(fileId, { include: ['XRefs', 'MediaInfo'] }, !!fileId);
+  const fileAniDbUrl = `https://anidb.net/file/${file?.AniDB?.ID}`;
+  const seriesShokoId = useMemo(() => get(file, 'SeriesIDs.0.SeriesID.ID', 0), [file]);
```
Comment (harshithmohan) [inline src/pages/utilities/FileSearch.tsx:166]:
```suggestion
  const seriesShokoId = file?.SeriesIDs?.[0].SeriesID.ID;
```
Memoization is costly if we are only accessing data from an object.

File: src/pages/utilities/FileSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+  const { data: seriesInfo } = useSeriesAniDBQuery(
+    get(file, 'SeriesIDs.0.SeriesID.AniDB', 0),
+    get(file, 'SeriesIDs.0.SeriesID.AniDB', 0) !== 0,
+  );
+  const { data: episodeInfo } = useEpisodeAniDBQuery(
+    get(file, 'SeriesIDs.0.EpisodeIDs.0.ID', 0),
+    get(file, 'SeriesIDs.0.EpisodeIDs.0.ID', 0) !== 0,
+  );
```
Comment (harshithmohan) [inline src/pages/utilities/FileSearch.tsx:174]:
```suggestion
  const seriesAnidbId = file?.SeriesIDs?.[0].SeriesID.AniDB;
  const episodeId = file?.SeriesIDs?.[0].EpisodeIDs[0].ID;
  const { data: seriesInfo } = useSeriesAniDBQuery(seriesAnidbId!, !!seriesAnidbId);
  const { data: episodeInfo } = useEpisodeAniDBQuery(episodeId!, !!episodeId);
```

Comment (harshithmohan) [reply src/pages/utilities/FileSearch.tsx:174]: More readable and less duplicate code

Comment (harshithmohan) [reply src/pages/utilities/FileSearch.tsx:174]: Also, need to change the type of `EpisodeIDs` to `EpisodeIDsType[]` in `XRefsType`. It seems to be wrong
image

File: src/pages/utilities/FileSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+      return prev - 1;
+    });
+  }, [selectedRows]);
+
+  const selectedId = useMemo(() => selectedRows[viewIndex]?.ID, [selectedRows, viewIndex]);
+
+  const columns = useMemo<UtilityHeaderType<FileType>[]>(
+    () => staticColumns,
+    [],
+  );
```
Comment (harshithmohan) [inline src/pages/utilities/FileSearch.tsx:339]: You can just pass `staticColumns` to the `UtilitiesTable` if you aren't going to add/remove anything. This is not needed

File: src/pages/utilities/FileSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+  const renderTotalFiles = () => (
```
Comment (harshithmohan) [inline src/pages/utilities/FileSearch.tsx:341]: We already have this as a component. Use that. `ItemCount.tsx`

File: src/pages/utilities/FileSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+    <div className="flex font-semibold">
+      <p className="text-panel-text-important">{fileCount ?? 0}</p>
+      &nbsp;
+      {fileCount <= 1 ? 'File' : 'Files'}
+    </div>
+  );
+
+  const renderMetadata = useMemo(() => {
```
Comment (harshithmohan) [inline src/pages/utilities/FileSearch.tsx:349]: This can be separated into a component

File: src/pages/utilities/FileSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+          />
+          <Menu
+            selectedRows={selectedRows}
+            setSelectedRows={setRowSelection}
+          />
+        </div>
+      </ShokoPanel>
+      <div className="flex grow justify-between overflow-y-auto">
+        <div className="flex w-full rounded-md border border-panel-border bg-panel-background p-8 md:max-w-[65%] lg:max-w-[75%]">
+          {(fileCount ?? 0) > 0
```
Comment (harshithmohan) [inline src/pages/utilities/FileSearch.tsx:397]:
```suggestion
          {fileCount > 0
```
`fileCount` is always defined

**Review by harshithmohan — CHANGES_REQUESTED** [review]

**Comment by harshithmohan** [comment]
https://github.com/ShokoAnime/Shoko-WebUI/assets/26010946/8b2f1675-8945-4f6a-a4c0-b09b5826409d

The scroll container changes when a file is selected 🤔

**Comment by harshithmohan** [comment]
image
File location is going out of the window

**Comment by harshithmohan** [comment]
image
Don't show this button if the anidb file id is not available. It's pointing to `https://anidb.net/file/undefined`

**Comment by harshithmohan** [comment]
`src/components/Utilities/Common/ItemCount.tsx`
This file can just be in `src/components/Utilities/`
We don't need a "common" folder. Everything in the "Utilities" folder will be common

**Review by harshithmohan — CHANGES_REQUESTED** [review]

**Review by harshithmohan — APPROVED** [review]

**Pull request merged**